### ISSUE TITLE
dirichlet constraints not thread safe

### ISSUE THREAD

**Issue by ftalbrecht** [pymor/dune-gdt#51]
It may happen for certain grids that we do not detect all dirichlet DoFs of an entity. This is true, for instance, for a triangular grid of the following form, where the bottom models the domain boundary
```
x----b----x
 \ 1 | 2 /
  \  |  /
   \ | /
  0 \|/ 3
-----a-----
 boundary
```
The dirichlet DoF associated with the vertex `a` is detected correctly on entities 0 and 3, but not on entities 1 and 2. This is due to the fact that we identify boundaries by looking at intersections and neither entity 1 nor 2 have intersection that are on the boundary.

The result is that there are certain combinations of global DoF indices, where there exist nontrivial entries in the system matrix that are not deleted during application of the dirichlet DoFs. This is for instance the case with the global index corresponding to vertices `a` and `b` (since `b` is neither present in entity 0 nor in 3).

The remedy within the current approach is to check the neighbors of an entity (or rather their intersections) too, when we are trying to determine if the vertex of an entity lies on the boundary. Since we have no way to determine the maximum number of entities associated with a vertex this is a nasty task and can only be satisfactory solved for some grids, on others we would need some kind of heuristics.

I am currently implementing such an approach that should be sufficient for the near future but we might have to look into other approaches of finding the DoFs associated with the boundary. We would not have this problem if we would create a global container of dirichlet DoFs, for instance, but we might have others.

And yes, @renemilk, this did show up in those tests that have been broken forever.

**Comment by ftalbrecht** [comment]
This is related to #44.

**Comment by ftalbrecht** [comment]
This is most likely the cause for #4.

**Comment by ftalbrecht** [comment]
64345a8 fixes the issue for the moment. I will leave this open until we find a satisfactory solution, however.

Though it is very unlikely, mind that **all your computations using CG FEM up to now might be invalid**, if you have used an `ALUGrid< 2, 2, consofming, simplex >` in a weird setup.

**Comment by ftalbrecht** [comment]
After the recent refactoring the `DirichletConstraints` class is just a container where the space adds dirichlet DoFs on each entity, which is just a forward to an `insert` call on a `std::set` member. I am not sure if that is thread safe. If not, each thread has to collect its dirichlet DoFs separately and we have to merge them afterwards. This can be nicely done in the assembler wrapper in the `finalize()` call.

**Comment by ftalbrecht** [comment]
Implemented in cad4519.
